WIP #18919 Align Java HTTP server DSL with Scala

jypma opened the pull request:
This replaces the Java HTTP server DSL with a Java-8 centric one which exposes all scala DSL concepts to be usable from Java, including custom directives, (un)marshallers, rejections, headers, and type safety for path and query parameters. 

Implemented according to the direction agreed upon with @ktoso in https://github.com/akka/akka/issues/18919 . The class `JavaRouteTest` demoes the Java DSL's capabilities.

List of things to be done (somewhat in order):
- [X] Migrated extra directives: `FileAndResourceDirectives`, `CodingDirectives`,`CookieDirectives`,`MiscDirectives`,  `HostDirectives`, `SchemeDirectives`, `WebSocketDirectives`, `CacheConditionDirectives`,  `RangeDirectives`
- [X] Migrate the existing tests in `akka-http-tests` once the DSL direction is agreed 
- [X] Fix compile errors in `akka-docs`
- [X] Fill in the missing directives that haven't been migrated yet 
- [x] Discuss how to resolve the introduced `TODO` and `FIXME` markers
- [ ] Replace the current Java `RouteTestKit` (which has its own implementation different from the scala `RouteTestKit`) with a [wrapper](https://github.com/jypma/scala-japi/blob/master/akkaStreamsTest/src/main/scala/com/tradeshift/scalajapi/akka/route/RouteTestKit.scala) that delegates and re-uses the scala one.
- [ ] Update textual documentation in `akka-docs`
- [x] MUST: Add `FormFieldDirectives`
- [x] MUST: Add `SecurityDirectives`, possibly Java-model a bit of the related classesupon
- [x] MUST: Add `DebuggingDirectives`
- [x] MUST: Add `FileUploadDirectives`
- [x] MUST: Add `RespondWithDirectives`

jypma commented:
@rkuhn or @drewhk, Konrad mentioned you could give it a quick peek in his absence. Maybe just to OK the general direction, while work is being done on the remaining directives. Thanks!

ktoso commented:
OK TO TEST

ktoso commented:
FYI @sirthias @jrudolph 
If you had a bit of time would be very nice if you could skim this new Java DSL proposal; The goal of being nearer to the Scala API also makes it easier for long term maintenance and feature parity between the APIs (which is why I find it compelling, and with JDK8 it doesn't look *that* bad).

It's not urgent, as DSLs stay experimental for now so we can keep working on it for now until we decide we like it.

I'll be back from vacation soon will dive into it properly, sorry for the timing on this one.

jypma commented:
(test failure expected, since the original `akka-http-tests` tests haven't been migrated yet)

rkuhn commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/RouteDirectives.scala:
```diff
@@ -0,0 +1,65 @@
+abstract class RouteDirectives extends RangeDirectives {
+  /**
+   * Java-specific call added so you can chain together multiple alternate routes using comma,
+   * rather than having to explicitly call route1.orElse(route2).orElse(route3).
```
good!

rkuhn commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/RouteDirectives.scala:
```diff
@@ -0,0 +1,65 @@
+    D.complete(ToResponseMarshallable(value)(marshaller.unwrap))
+  }
+
+  //FIXME look at the other occurrences of java.lang.Iterable, they might better be varargs at the end?
```
yes, that is a good point

rkuhn commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/PathDirectives.scala:
```diff
@@ -1,187 +1,59 @@
+
+/**
+ * Only path prefixes are matched by these methods, since any kind of chaining path extractions
+ * in Java would just look cumbersome without operator overloads; hence, no PathMatcher for Java.
```
I think this is a viable way around the overloading boilerplate madness that would otherwise ensue. It makes Java routes a bit more verbose than their Scala equivalent, but that cost is due to the host language.

Users should be able to compose and reuse these, right?

jypma replied:
Definitely, we also use them as such. For example, you could have a method
```
Route standardPath(BiFunction[UUID,String,Route] inner) {
  return path("documents", () -> 
    path(UUID(), id -> 
      path("variants", () -> 
        path(variant -> inner.apply(id, variant)
      )
    )
  )
}
```

rkuhn replied:
Great, we should definitely take care to document these composition patterns!

rkuhn commented:
Sorry for the delay, I think this looks very promising! I have not had the time to study the actual implementation, though.

jypma commented:
@rkuhn Thanks for the comments so far! Do you agree enough to the general direction to warrant the time investment of starting to add some of the missing directives, and migrating some tests?

rkuhn commented:
Yes, as I said I think this is looking good—I just didn’t realize that I was reading the actual implementation (i.e. I didn’t even take in the full simplicity of the proposal).

If I understand correctly the difference between Java and Scala will then only be that Scala directives can be composed directly while Java directives can only be composed by applying them and wrapping that application in a lambda. This difference is due to the different language capabilities, so I think it is a fine trade-off to make.

@drewhk @johanandren please review as well

johanandren commented:
I've looked through the changes, I don't have any detail feedback yet, but I think it looks really promising. I feel this is a big improvement for the Java side of the API.

johanandren commented:
Spent a bit of time hands-on with it. Some things are missing, but that's to expect, My opinion is that this feels really nice to build routes with. I'm all :+1:, move forward with this.

johanandren commented:
Hmmm. Unrelated incremental compilation failure, as far as I can see.

johanandren commented:
PLS BUILD

johanandren commented:
Haha, sorry, I failed to context switch and thought this was something I just pushed on another PR.

rkuhn commented on akka-http/src/main/scala/akka/http/javadsl/server/Directives.scala:
```diff
@@ -4,22 +4,18 @@
-    }
+
+  // These are repeated here since sometimes (?) the Scala compiler won't actually generate java-compatible
+  // signatures for varargs methods, making them show up as Seq<Object> instead of T... in Java.
```
Yes, indeed, this is a bug I have run into with genjavadoc as well. I hope this work-around is reliable.

rkuhn commented on akka-http/src/main/scala/akka/http/javadsl/server/JavaScalaTypeEquivalence.scala:
```diff
@@ -0,0 +1,63 @@
+  // (being the scaladsl one)
+
+  // These are gathered here rather than doing explicit asInstanceOf throughout the code, in order
+  // to get predictable compile errors if this hierarchy might ever change.
```
:+1:

jypma replied:
@ktoso I recall finding a similar class in the existing java/scala DSL that adds implicit `toScala` / `toJava` methods instead. I opted for direct implicit conversions here, since very often in the DSL it's not just a`T` that needs to be converted, but rather a `Seq[T]`, `Future[T]`, `Flow[T]` etc. Still, it's probably something that should be aligned.

rkuhn commented on akka-http-tests/src/test/java/akka/http/javadsl/server/values/FormFieldsTest.java:
```diff
@@ -98,227 +99,18 @@ public void testByteFormFieldExtraction() {
     }
 
     @Test
-    public void testShortFormFieldExtraction() {
```
will these tests return or have they been moved elsewhere?

jypma replied:
The type-specific tests were dropped since the new `FormFieldDirectives` only has a single typed `formField` method, that takes an `Unmarshaller[String,T]`. So `BYTE()` above covers that. We can bring them back for consistency of course, if you prefer.

The other unmarshallers in `StringUnmarshallers` are tested in `ParametersTest`.

rkuhn commented:
This is a really heavy chunk of work, but what I see looks very good. I guess we should also include automated tests that compare the javadsl and scaladsl packages to make sure that everything is covered—just like we have for the Streams DSLs.

I think we should make this the main feature for Akka 2.4.3, we’ll have to collaborate to make that happen, though. We should probably make your branch the center for this action if you have the time to merge PRs against it—we could also move this branch into the akka/akka repo so that we can merge things.

In any case there is no hurry, we’ll need to get 2.4.2 out first (hopefully on Wednesday).

jypma commented:
PR's to this branch is OK with me; in that case, what would the process be once we start squashing commits for release, especially w.r.t. authors?

rkuhn commented:
We either don’t squash (in case of sizable contributions that are not from you) or you get all the credit, but let’s see once we get there :-)

jypma commented:
Alright. Either way, I'm keeping the shopping list in the description up to date. Currently poking at `akka-docs` (the code part). I expect there's some actual doc rewriting to be done as a part of this, at some point.

ktoso commented on akka-http-marshallers-java/akka-http-jackson/src/main/java/akka/http/javadsl/marshallers/jackson/Jackson.java:
```diff
@@ -0,0 +1,55 @@
+		} catch (IOException e) {
+			throw new IllegalArgumentException("Cannot unmarshal JSON as " + expectedType.getSimpleName(), e);
+		}
+	}
```
Why not in the previously existing class in this module? Seems to do the same

ktoso replied:
So it's likely for easier calling into this class from Java

ktoso replied:
Please add copyright header then :)

jypma replied:
I assumed there was some script-foo for the copyright headers. Will add.

ktoso commented on akka-http-testkit/src/main/scala/akka/http/javadsl/testkit/TestRouteResult.scala:
```diff
@@ -151,16 +204,20 @@ abstract class TestResponse(_response: HttpResponse, awaitAtMost: FiniteDuration
+  protected def assertEqualsKind(expected: Int, actual: Int, kind: String): TestRouteResult = {
     assertEquals(expected, actual, s"Unexpected $kind!")
     this
   }
```
nice, and good comments

ktoso commented on akka-http-tests/src/main/java/akka/http/javadsl/server/examples/petstore/PetStoreExample.java:
```diff
@@ -6,46 +6,55 @@
-          delete(handleReflectively(controller, "deletePet", petId))
+            // 2. using a method
+            put(() -> 
+              entityAs(Jackson.jsonAs(Pet.class), thePet -> 
```
potential naming bikeshed here? The scala one reads nicely because `entity(as...)`

jypma replied:
Maybe `entityUsing`? I'll check if we can overload just `entity`.

jypma replied:
Here are a few alternatives:
1. `entity(Jackson.unmarshaller(Pet.class), thePet -> ...)`
2. `entity(Jackson.fromJson(Pet.class), thePet -> ...)`
3. `entityUsing(Jackson.unmarshaller(Pet.class), thePet -> ...)`
4. `entityUsing(Jackson.to(Pet.class), thePet -> ...)`

@ktoso which do you like best?

I'll rename the marshalling part of `Jackson` appropriately to what we choose here.

In all cases, I think I'll drop the variant that takes a `Class` as first argument. It's leftover from Scala where an unmarshaller is pulled in implicitly by type. Dropping it simplifies the DSL, and I think it's good to have a word ("json", "xml", "xmlOrJson") between the type and the http entity.

rkuhn replied:
I like the first one: the pattern should be that you can stick something that’s called “unmarshaller” into the `entity` directive.

jypma replied:
Alright, going with (1).

ktoso commented on akka-http-tests/src/main/java/akka/http/javadsl/server/examples/simple/SimpleServerApp.java:
```diff
@@ -4,84 +4,76 @@
-                int result = xVal + yVal;
-                return ctx.complete(String.format("%d + %d = %d", xVal, yVal, result));
-            }
+        Route addHandler = param(INT(), "x", x ->
```
looks a bit weird, in Java INT would make me assume its a constant, yet it seems to be a function?

rkuhn replied:
no, as commented in a TODO somewhere above this is supposed to be an enum (once rewritten in Java)

ktoso replied:
I see, didnt find that TODO yet. enum makes sense

jypma replied:
They're actually intended to become `public static final`s in a class. So then it'll just be `INT`. But I might just make this one `INTEGER`, since that's what the boxed class is called in Java.

ktoso commented on akka-http-tests/src/main/java/akka/http/javadsl/server/examples/simple/SimpleServerApp.java:
```diff
@@ -4,84 +4,76 @@
+                path("add", () -> addHandler),
+                path("subtract", () ->
+                    param(INT(), "x", x ->
+                        param(INT(), "y", y ->
```
I wonder if we could push it even more `param2(named("x").asInteger(), named("y").asDouble(), (x, y) ->` etc

jypma replied:
possible, but with `param2` you wouldn't even need the extra `named` calls, just `param2(INT, "x", INT, "y", (x,y) -> ...`. 

I wouldn't use it past 2 though: it pulls the argument order too far away from the place where they're declared. Should I add param2?

ktoso commented on akka-http-tests/src/main/java/akka/http/javadsl/server/examples/simple/SimpleServerApp.java:
```diff
@@ -4,84 +4,76 @@
-                    // bind handler by reflection
-                    handleReflectively(SimpleServerApp.class, "multiply", xSegment, ySegment)
+                path(segment("multiply").slash(IntegerSegment()).slash(IntegerSegment()), 
+                    this::multiply
```
very nice

ktoso commented on akka-http-tests/src/test/java/akka/http/javadsl/server/JavaRouteTest.java:
```diff
@@ -0,0 +1,351 @@
+package akka.http.javadsl.server;
```
missing license header

jypma replied:
adding

ktoso commented on akka-http-tests/src/test/java/akka/http/javadsl/server/JavaRouteTest.java:
```diff
@@ -0,0 +1,351 @@
+    
+    public Route getRoute() {
+        return route(
+            path(segment("hello").slash("world"), () ->
```
Mixed feelings on `segment("hello").slash("world")`, makes it sound like world is some kind of slash thingy, where as slash is only the separator... How about `segment("hello").slash().segment("world")` even if slighly more verbose, it reads more correctly

rkuhn replied:
Nah, that’s too verbose; I didn’t have this reaction when first seeing this.

jypma replied:
isn't that already possible? `segment("hello").slash(segment("world"))`.  The `String` variants are just for convenience.

ktoso commented on akka-http-tests/src/test/java/akka/http/javadsl/server/directives/PathDirectivesTest.java:
```diff
@@ -58,12 +63,20 @@ public void testRawPathPrefix() {
                 rawPathPrefix(PathMatchers.SLASH(), "pet", PathMatchers.SLASH(), "", PathMatchers.SLASH(), "cat", PathMatchers.END()).route(
                     complete("The cat!")
                 )
             );
```
comment can be removed AFAICS

ktoso commented on akka-http-tests/src/test/java/akka/http/javadsl/server/directives/PathDirectivesTest.java:
```diff
@@ -294,11 +296,9 @@ public void testHexLongMatcher() {
         TestRoute route =
             testRoute(
-                path("pets", rest).route(completeWithValueToString(rest))
+                path(Rest(), remainingPath -> complete(remainingPath))
```
I saw a proposal somewhere while I was sick to rename `Rest` to `Remaining`, I'd like that :)

jypma replied:
Will do.

2beaucoup commented on akka-http-tests/src/test/java/akka/http/javadsl/server/JavaRouteTest.java:
```diff
@@ -0,0 +1,351 @@
+import akka.japi.pf.PFBuilder;
+import akka.util.ByteString;
+import static akka.http.javadsl.server.StringUnmarshaller.*;
+import static akka.http.javadsl.server.PathMatchers.*;
```
I'd vote against the usage of static imports. One disadvantage is that I can only add the *after* I've discovered the methods (via auto complete). My bet is that only a minority of users is using them. Let's see how the API looks to the others.

WDYT?

ktoso replied:
Yeah, that is a good point - let's try to make it such that one can `extends HttpService` or something to get everything you need. The Directives were attempting that previously already, so it's about hooking into there with the matchers as well.

jypma replied:
Extending a class is an OK way to get started and for quick docs and tests, but my experience with static imports is actually the opposite of what is being stated. Getting to constants by extending an interface/base class is also explicitly mentioned as an [anti-pattern](https://docs.oracle.com/javase/1.5.0/docs/guide/language/static-import.html) even by Sun. Remember, most of `StringUnmarshallers` and `PathMatchers` will become "just" `public static final`s.

 Having to extend a class to get to constants has two big drawbacks:
- A class can only extend 1 base class. Requiring this makes it impossible to mix akka with other frameworks.
- All of the inherited constants appear on the public API of the class being written (unless we make them protected, completely removing the ability to statically import)

I'll refactor the aforementioned classes into constants first, let's see what it looks like then. Currently people also don't need to extend `StatusCodes` to get to the HTTP status codes either.

rkuhn replied:
I’m with Jan on this one, without mixins we cannot apply the Scala solution to Java.

rkuhn replied:
And the Scala solution is also inherently limiting, see the difficulty of adding anything to a trait as exemplified in #19746. We should be extremely careful with these things.

2beaucoup replied:
I'm not against static methods. I just wouldn't import them statically in tests/docs. ;)

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/HttpService.scala:
```diff
@@ -38,8 +37,8 @@ trait HttpServiceBase {
-    val r: server.Route = RouteImplementation(route)
-    Http(system).bind(interface, port).toMat(Sink.foreach(_.handleWith(RouteResult.route2HandlerFlow(r))))(Keep.left).run()(materializer).toJava
+    val r: server.Route = route.toScala
+    Http(system).bind(interface, port).toMat(Sink.foreach(_.handleWith(r)))(Keep.left).run()(materializer).toJava
```
good

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/JavaScalaTypeEquivalence.scala:
```diff
@@ -0,0 +1,63 @@
+object JavaScalaTypeEquivalence {
+  // We define implicit conversions from the javadsl model objects to the scaladsl subclasses,
+  // for all types where there is only 1 valid direct subclass of the javadsl type
+  // (being the scaladsl one)
```
As you noticed we have such concept in the form of `JavaMapping` so I'd like to keep those there too.
It's via implicits too there so we can simply provide an additional import there to enable conversions instead of asJava/asScala. I can help out with that for example.

jypma replied:
Good idea. Let's do it on the next run after I fix the lower-hanging comments. And maybe after the thing _really_ compiles (I totally missed `akka-docs` last time I claimed that...)

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/Marshaller.scala:
```diff
@@ -4,8 +4,108 @@
+  }
+}
+
+case class Marshaller[A,B] private (implicit val unwrap: marshalling.Marshaller[A,B]) {
```
final

ktoso replied:
Seems a bit weird... where is this used why `implicit unwrap`?

jypma replied:
It's the way to create a marshaller from Java, using the factory methods in the companion object. `implicit` can probably go away, I think it's a leftover from when I moved the implicit conversions to the single class.

ktoso replied:
Removal of implicit sounds good :+1:

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/Marshaller.scala:
```diff
@@ -4,8 +4,108 @@
+    wrap(scaladsl.marshalling.Marshaller.byteStringMarshaller(t))
+  }
+  
+  def opaque[A,B](f: function.Function[A,B]): Marshaller[A, B] = {
```
`combined` / `strict` seem missing, should be here too right?

ktoso replied:
Likely also the factories `withOpenCharset` and `withFixedContentType`

jypma replied:
Yupp. I've only been adding what we needed for creating custom XML / JSON marshallers, and the String test stuff, so far. Rest will be added.

ktoso replied:
Okey, excellent :) Sorry if I sounded nagging, trying to not miss on anything by accident :+1:

jypma replied:
Actually I left out `strict` on purpose since that would require a Java API for the `Marshalling` trait. I haven't found a need for it... we've built a `Accept`-aware JSON+XML marshaller by simply doing `oneOf(xml, json)`. Should we pull adding `strict` into scope?

And isn't `combine` covered by `instance.map()` / `instance.compose()`, given that there are no implicits in Java anyways?

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/Marshaller.scala:
```diff
@@ -4,8 +4,108 @@
+import JavaScalaTypeEquivalence._
+
+object Marshaller {
+  def wrap[A,B](scalaMarshaller: marshalling.Marshaller[A,B]) = Marshaller()(scalaMarshaller)
```
Wrap and unwrap sound like sidestepping the asJava/asScala infra we have in JavaMapping.
On the other hand wrap/unwrap I see may be intended for users to use when given "the wrong marshaller"?

jypma replied:
Sure, those names can be aligned. Speaking of which, the existing conversions also say `asJava` as well as `toJava` (java8-compat, collection convertors I believe).

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/PathMatcher.scala:
```diff
@@ -0,0 +1,110 @@
+package akka.http.javadsl.server
```
license header

jypma replied:
added.

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/PathMatcher.scala:
```diff
@@ -0,0 +1,110 @@
+  def wrap0(scalaMatcher: SPathMatcher[Unit]) = new PathMatcher0(scalaMatcher)
+  def wrap1[T](scalaMatcher: SPathMatcher[Tuple1[T]]) = new PathMatcher1(scalaMatcher)
+  def wrap2[T1, T2](scalaMatcher: SPathMatcher[(T1, T2)]) = new PathMatcher2(scalaMatcher)
+}
```
looks good

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/RejectionHandler.scala:
```diff
@@ -1,190 +1,57 @@
+  def defaultHandler = RejectionHandler(server.RejectionHandler.default)
+}
 
+case class RejectionHandler(unwrap: server.RejectionHandler) {
```
does not need to be a case class, also, likely can be `final`. (anything case class is harder to evolve in bin compat way, so let's try to not use them where it's not a datatype)

jypma replied:
Done. also applied `asScala`, and did the same to `ExceptionHandler`.

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/Rejections.scala:
```diff
@@ -0,0 +1,89 @@
+package akka.http.javadsl.server
```
please add license header

jypma replied:
added.

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/RequestContext.scala:
```diff
@@ -1,90 +1,8 @@
-/*
- * Copyright (C) 2009-2016 Typesafe Inc. <http://www.typesafe.com>
- */
-
```
keep that header please :)

jypma replied:
class is gone now.

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/RequestContext.scala:
```diff
@@ -1,90 +1,8 @@
-  def reject(customRejection: CustomRejection): RouteResult
+//TODO do we still need this? all routes for Java are scala wrapped routes anyways
+case class RequestContext(toScala: akka.http.scaladsl.server.RequestContext) {
+  def getRequest: HttpRequest = toScala.request
```
the Request type here I'd want to be the javadsl model HttpRequest, so I'd say we still need this java context type

ktoso replied:
To put it in some different light: when a java user suddenly has to reach into scaladsl types we messed up :)

jypma replied:
I removed the Java `RequestContext` and its extractor. Everything delegates down into the Scala request components directly anyways, and there are extractors for that. And not everything in the Scala RequestContext has a Java model representation. So in my opinion, we can leave that for a later step.

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/Route.scala:
```diff
@@ -4,11 +4,40 @@
-trait Route
\ No newline at end of file
+trait Route {
+  private[http] def toScala: scaladsl.server.Route
```
needs `/** INTERNAL API */` comment when has `private[http]` (as it appears as public for Java classes)

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/StringUnmarshaller.scala:
```diff
@@ -0,0 +1,32 @@
+  def async[B](f: java.util.function.Function[String, CompletionStage[B]]): Unmarshaller[String, B] = Unmarshaller.async(f)
+
+  def sync[B](f: java.util.function.Function[String, B]): Unmarshaller[String, B] = Unmarshaller.sync(f)
+}
```
let's not forget to make these into a Java class then (add FIXME until done or tracked somewhere else already)?

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/Unmarshaller.scala:
```diff
@@ -4,11 +4,95 @@
-trait Unmarshaller[T] {
-  def classTag: ClassTag[T]
+object Unmarshaller {
+  private implicit val _ = javaToScalaHttpEntity
```
somewhat cryptic, can we give it a proper identifier?

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/Unmarshaller.scala:
```diff
@@ -4,11 +4,95 @@
+    
+}
+
+case class Unmarshaller[A,B] private (implicit val unwrap: unmarshalling.Unmarshaller[A,B]) {
```
Implicit seems not needed. Naming we'll want to revisit

jypma replied:
added comment on why it's implicit. Also turned into `final class` instead of `case class`, and renamed `unwrap` to `asScala`.

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/CacheConditionDirectives.scala:
```diff
@@ -38,9 +37,9 @@ abstract class CacheConditionDirectives extends BasicDirectives {
-    RouteStructure.Conditional(lastModified = Some(lastModified))(innerRoute, moreInnerRoutes.toList)
+  //@varargs
+  //def conditional(lastModified: DateTime, innerRoute: Route, moreInnerRoutes: Route*): Route =
+  //  RouteStructure.Conditional(lastModified = Some(lastModified))(innerRoute, moreInnerRoutes.toList)
```
let's put a FIXME so we don't leave those lying around accidentally

jypma replied:
There's an explicit TODO item in the top of the PR for these, I promise I won't forget :)

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/CodingDirectives.scala:
```diff
@@ -5,42 +5,84 @@
    */
-  @varargs def encodeResponse(coders: Coder*): Directive =
-    Directives.custom(RouteStructure.EncodeResponse(coders.toList))
+  // TODO provide Java API for creating Encoder, by extracting a Java interface from the Scala trait
```
For that we usually have the java interface, and make the scala one extend it (you'll notice in many other classes in HTTP)

jypma replied:
I considered that, but wouldn't that break binary compatibility in `-core`?

ktoso replied:
No why, adding methods there should be fine.

jypma replied:
Looks like there is quite a bit of stuff in there. Since it's not possible to create a custom `Encoder` with the current DSL either, could we postpone this to a future PR?

ktoso replied:
Okey, let's postopne and add in a separate PR, I made a ticket for it https://github.com/akka/akka/issues/19882

ktoso replied:
(Please add the ticket number to the todo comment )

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/FormFieldDirectives.scala:
```diff
@@ -0,0 +1,72 @@
+  }
+
+  def formFieldList[T](t: Unmarshaller[String, T], name: String, inner: JFunction[java.util.List[T], Route]): Route = {
+    import t.unwrap
```
I see why making it implicit was useful now... perhaps harmless and we can keep it so.

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/HeaderDirectives.scala:
```diff
@@ -0,0 +1,70 @@
+ * (copy / refer to scala DSL doc here)
+ * 
+ * FIXME When implementing custom headers in Java, just extend akka.http.scaladsl.model.headers.CustomHeader,
+ * since the java API's CustomHeader class does not implement render().
```
Yeah that's something we'll need to fix then.

ktoso commented on akka-http/src/main/scala/scala/util/matching/JavaConverters.scala:
```diff
@@ -0,0 +1,10 @@
+package scala.util.matching
```
we can't cross proper package boundaries :-) Will have to live in akka.http.something.something :) AFAIR we (sadly) so have a util class somewhere, you could put it there)

jypma replied:
@ktoso See the comment. I added this in order to turn a Java `Pattern` into a Scala `Regex` without recompiling it. That's pretty important, since many directives will be invoked from within lambdas that are called on every route, and you don't want to do the conversion + recompilation inside the directive for every call. 

But scala's `Regex` constructor is package-private, so that's why I put it in this package. I know it's a hack, and I'm open for alternatives that give the same performance without sacrificing DSL comfort level. Maybe adding a directive to the Scala DSL that takes a plain Java `Pattern`?

ktoso replied:
Comment did not explain it's package private :)
Sadly this will break osgi packaging and people will come at us with pitchforks and fire ;)
We should be able to achieve the same thing with some reflection, to avoid the package private problem them.

jypma replied:
Actually, package-private in Scala means public in Java; maybe I should just write the regex wrapper in Java, calling into `new Regex(thePattern, EMPTY_SEQ)`, and then use that in the directive? Seems a bit cleaner than reflection, and at least we have _one_ compiler future-checking it.

ktoso commented:
Went through once, very promising indeed! Let's figure out after 2.4.2 how to split the remaining work such that it's polished and we can ship it soon. :)

vasseurjb commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/PathDirectives.scala:
```diff
@@ -136,52 +66,213 @@ abstract class PathDirectives extends MiscDirectives {
+    D.rawPathPrefix(pm.toScala) { inner.get().toScala }
+  }
+  def rawPathPrefix[T1](pm: PathMatcher1[T1], inner: Function[T1, Route]): Route = ScalaRoute {
+    D.rawPathPrefix(pm.toScala) { t1 ⇒ inner.apply(t1).toScala }
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/vasseurjb/akka/file/2046270054/issues/source?bid=3132306&fileBranchId=3132379#l136)

vasseurjb replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/vasseurjb/akka/file/2046270054/issues/source?bid=3132306&fileBranchId=3132379#l136)

vasseurjb commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/PathDirectives.scala:
```diff
@@ -136,52 +66,213 @@ abstract class PathDirectives extends MiscDirectives {
+    D.pathPrefix(segment) { inner.get.toScala }
+  }
+  def pathPrefix(inner: java.util.function.Function[String, Route]): Route = ScalaRoute {
+    D.pathPrefix(PathMatchers.Segment) { element ⇒ inner.apply(element).toScala }
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/vasseurjb/akka/file/2046270054/issues/source?bid=3132306&fileBranchId=3132379#l111)

vasseurjb replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/vasseurjb/akka/file/2046270054/issues/source?bid=3132306&fileBranchId=3132379#l111)

vasseurjb commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/PathDirectives.scala:
```diff
@@ -136,52 +66,213 @@ abstract class PathDirectives extends MiscDirectives {
-    RedirectToTrailingSlashIfMissing(redirectionStatusCode)(innerRoute, moreInnerRoutes.toList)
+  def redirectToTrailingSlashIfMissing(redirectionType: StatusCode, inner: Supplier[Route]): Route = ScalaRoute {
+    redirectionType match {
+      case r: Redirection ⇒ D.redirectToTrailingSlashIfMissing(r) { inner.get().toScala }
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/vasseurjb/akka/file/2046270054/issues/source?bid=3132306&fileBranchId=3132379#l225)

vasseurjb replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/vasseurjb/akka/file/2046270054/issues/source?bid=3132306&fileBranchId=3132379#l225)

vasseurjb commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/PathDirectives.scala:
```diff
@@ -1,112 +1,42 @@
+ * Just nest path() directives with the required types, ending in pathEnd() if you want to fail
+ * further paths.
+ */
+abstract class PathDirectives extends ParameterDirectives {
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Number of methods in class exceeds 30](https://www.codacy.com/app/vasseurjb/akka/file/2046270054/issues/source?bid=3132306&fileBranchId=3132379#l31)

vasseurjb replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Number of methods in class exceeds 30](https://www.codacy.com/app/vasseurjb/akka/file/2046270054/issues/source?bid=3132306&fileBranchId=3132379#l31)

vasseurjb commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/PathDirectives.scala:
```diff
@@ -136,52 +66,213 @@ abstract class PathDirectives extends MiscDirectives {
-      case x                 ⇒ throw new IllegalArgumentException(s"Matcher of class ${x.getClass} is unsupported for PathDirectives")
-    }
+  private def unmarshal[T](t: Unmarshaller[String, T], inner: java.util.function.Function[T, Route]) = { element: String ⇒
+    D.extractRequestContext { ctx ⇒
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/vasseurjb/akka/file/2046270054/issues/source?bid=3132306&fileBranchId=3132379#l263)

vasseurjb replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/vasseurjb/akka/file/2046270054/issues/source?bid=3132306&fileBranchId=3132379#l263)

vasseurjb commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/PathDirectives.scala:
```diff
@@ -136,52 +66,213 @@ abstract class PathDirectives extends MiscDirectives {
+    D.path(p.toScala) { inner.get.toScala }
+  }
+  def path[T](p: PathMatcher1[T], inner: JFunction[T, Route]): Route = ScalaRoute {
+    D.path(p.toScala) { t1 ⇒ inner.apply(t1).toScala }
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/vasseurjb/akka/file/2046270054/issues/source?bid=3132306&fileBranchId=3132379#l96)

vasseurjb replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/vasseurjb/akka/file/2046270054/issues/source?bid=3132306&fileBranchId=3132379#l96)

vasseurjb commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/PathDirectives.scala:
```diff
@@ -136,52 +66,213 @@ abstract class PathDirectives extends MiscDirectives {
+    D.pathPrefix(p.toScala) { t1 ⇒ inner.apply(t1).toScala }
+  }
+  def pathPrefix[T1, T2](p: PathMatcher2[T1, T2], inner: BiFunction[T1, T2, Route]): Route = ScalaRoute {
+    D.pathPrefix(p.toScala) { (t1, t2) ⇒ inner.apply(t1, t2).toScala }
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/vasseurjb/akka/file/2046270054/issues/source?bid=3132306&fileBranchId=3132379#l120)

vasseurjb replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/vasseurjb/akka/file/2046270054/issues/source?bid=3132306&fileBranchId=3132379#l120)

vasseurjb commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/PathDirectives.scala:
```diff
@@ -136,52 +66,213 @@ abstract class PathDirectives extends MiscDirectives {
+    D.pathPrefix(p.toScala) { inner.get.toScala }
+  }
+  def pathPrefix[T](p: PathMatcher1[T], inner: JFunction[T, Route]): Route = ScalaRoute {
+    D.pathPrefix(p.toScala) { t1 ⇒ inner.apply(t1).toScala }
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/vasseurjb/akka/file/2046270054/issues/source?bid=3132306&fileBranchId=3132379#l117)

vasseurjb replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/vasseurjb/akka/file/2046270054/issues/source?bid=3132306&fileBranchId=3132379#l117)

vasseurjb commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/PathDirectives.scala:
```diff
@@ -136,52 +66,213 @@ abstract class PathDirectives extends MiscDirectives {
+    D.pathSuffixTest(pm.toScala) { inner.get().toScala }
+  }
+  def pathSuffixTest[T1](pm: PathMatcher1[T1], inner: Function[T1, Route]): Route = ScalaRoute {
+    D.pathSuffixTest(pm.toScala) { t1 ⇒ inner.apply(t1).toScala }
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/vasseurjb/akka/file/2046270054/issues/source?bid=3132306&fileBranchId=3132379#l210)

vasseurjb replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/vasseurjb/akka/file/2046270054/issues/source?bid=3132306&fileBranchId=3132379#l210)

vasseurjb commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/PathDirectives.scala:
```diff
@@ -136,52 +66,213 @@ abstract class PathDirectives extends MiscDirectives {
+    D.rawPathPrefixTest(pm.toScala) { t1 ⇒ inner.apply(t1).toScala }
+  }
+  def rawPathPrefixTest[T1, T2](pm: PathMatcher2[T1, T2], inner: BiFunction[T1, T2, Route]): Route = ScalaRoute {
+    D.rawPathPrefixTest(pm.toScala) { case (t1, t2) ⇒ inner.apply(t1, t2).toScala }
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/vasseurjb/akka/file/2046270054/issues/source?bid=3132306&fileBranchId=3132379#l174)

vasseurjb replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/vasseurjb/akka/file/2046270054/issues/source?bid=3132306&fileBranchId=3132379#l174)

ktoso commented:
wat, why codacy suddenly decided to go out and comment on things... I completely removed it, sorry for the noise.

wjur commented:
A side note: this PR, as it adds the missing formField Java directives, it fixes #19519. You might want to add the id into the title (or do anything else that's appropriate).

rkuhn commented:
Small update: this branch will most definitely be merged eventually, but we cannot merge it before it is “finished” in the sense that `master` cannot ever have a broken javadsl—this means that until it is complete enough to be considered for potential release we will have to work against this branch, quite possibly in collaboration (as discussed above).

jypma commented:
@rkuhn Thanks for the update. What specific parts are you considering broken at the moment? To me only the documenation update is what's left; all other open TODO items are extensions to the current Java DSL's possibilities.

In that context, could you or @ktoso take a few minutes to provide comments on the open questions and discussion points in this PR? I feel they should be aligned before any documentation is updated. Thanks.

rkuhn commented:
Okay, I didn’t check in depth and assumed that some TODOs and especially FIXMEs meant that it is not yet fully usable. If the current state covers all that was previously possible, then we can merge as soon as no “wrong” docs are around any longer. Porting the Scala quickstart guide to Java will then be a separate ticket. I’ll review the FIXMEs after having finished with the rest of the PRs (hopefully still today, but possibly tomorrow). Concerning the route testkit I’ll have to take a look at the current state, this might want to be a separate ticket in any case.

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/FileAndResourceDirectives.scala:
```diff
@@ -17,112 +25,229 @@ trait ContentTypeResolver {
- */
-case class StaticContentTypeResolver(contentType: ContentType) extends ContentTypeResolver {
-  def resolve(fileName: String): ContentType = contentType
+// TODO Javify DirectoryListing
```
that one we just need to do AFAICS :)

jypma replied:
will do

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/FileAndResourceDirectives.scala:
```diff
@@ -17,112 +25,229 @@ trait ContentTypeResolver {
+      override def apply(fileName: String) = resolver.resolve(fileName)
+    }
+
+  private[FileAndResourceDirectives] implicit def scalaRenderer(renderer: DirectoryRenderer) =
```
- `private[http]` is enough
- `private[X]` does not affect Java, in fact it makes it `public` in Java (sic), so all cases where we do that need a comment with `/** INTERNAL API */` (so people can't complain if we remove them - it is internal api

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/HeaderDirectives.scala:
```diff
@@ -0,0 +1,105 @@
+ * 
+ * 
+ * FIXME When implementing custom headers in Java, just extend akka.http.scaladsl.model.headers.CustomHeader,
+ * since the java API's CustomHeader class does not implement render().
```
Ouch! That's an omission I'd say.
It seems simply adding `  final def render[R <: Rendering](r: R): r.type = r ~~ name ~~ ':' ~~ ' ' ~~ value`
inside java's CustomHeader should bin-compat right?

(reasoning - we can't force java people to "hey you need to extend scaladsl")

ktoso replied:
Sorry, I'm blind - we can't because of the path dependent type in the render signature... 
/me thinking about it (could be solved after the initial PR IMO)

rkuhn replied:
That’s not a path-dependent type, in Java this can just return type `R`.

ktoso commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/HeaderDirectives.scala:
```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
```
Thanks for applying that already :)

ktoso commented on akka-http-tests/src/test/java/docs/http/javadsl/server/HandlerExampleDocTest.java:
```diff
@@ -248,43 +146,56 @@ public void testDeferredResultAsyncHandler() {
         //#async-service-definition
 
+        // FIXME rewrite this beyond just fixing compiler errors, so it actually looks nice.
+        // e.g. extract getting the x and y param into a reusable method, like requestvals did before.
```
It's not too bad actually, fix if you have time but not a blocker

rkuhn replied:
This could be a TODO to inspire documentation improvements, which should now probably also synchronized between Scala and Java.

jypma replied:
cleaned this up a bit

ktoso commented on akka-http-tests/src/test/java/akka/http/javadsl/server/JavaRouteTest.java:
```diff
@@ -0,0 +1,342 @@
+import akka.http.javadsl.model.headers.Accept;
+import akka.http.javadsl.model.headers.RawHeader;
+import akka.http.javadsl.testkit.JUnitRouteTest;
+//FIXME discuss how to provide a javadsl.CustomHeader where render() is either pre-implemented or trivial to write in Java
```
I'll play around with some ideas on your branch now

ktoso commented:
> COULD: Add DebuggingDirectives, FileUploadDirectives, RespondWithDirectives

Java DSL should have feature parity so IMHO that's a MUST (I can help if you're stuck, but we need them before we announce the new one)

ktoso commented:
I looked at https://github.com/jypma/scala-japi/blob/master/akkaStreamsTest/src/main/scala/com/tradeshift/scalajapi/akka/route/RouteTestKit.scala and it's usage and it looks good, let's include in this PR. 

I'm not sure about the `on` word used in the tests, hmmm what could be a better name there, perhaps just `routeTest`?

rkuhn commented on akka-http/src/main/scala/akka/http/javadsl/server/ExceptionHandler.scala:
```diff
@@ -4,6 +4,13 @@
-  def handle(exception: RuntimeException): Route
+import akka.http.scaladsl.server
+
+// TODO give this a real builder to save one extra set of parens to PFBuilder
```
yes, please move from TODO to ticket

jypma replied:
just created the builder instead.

rkuhn commented on akka-http/src/main/scala/akka/http/javadsl/server/directives/ScalaRoute.scala:
```diff
@@ -0,0 +1,39 @@
+    case ScalaRoute(altRoute) ⇒
+      ScalaRoute(toScala ~ altRoute)
+
+    case _ ⇒ throw new IllegalArgumentException("TODO wrap java route in scala")
```
I don’t fully understand this exception message: isn’t it enough to just not unwrap the ScalaRoute in this case?

jypma replied:
I removed the clause instead. It's no longer intended to have alternate user Java implementations of `Route`.

rkuhn commented on akka-http-tests/src/test/java/akka/http/javadsl/server/values/HeadersTest.java:
```diff
@@ -37,21 +33,24 @@ public void testValueByName() {
     }
+    
+    /*
+    FIXME A replacement of the "instance" RequestVal feature would probably be to allow easy creation of a custom header in Java DSL, and extract -> re-apply that instead.
```
I don’t understand this one; either explain better or fix it :-)

jypma replied:
I'm removing the test instead. I initially thought that the test's `.instance()` here was referring to an "instance" of a custom modelled Java header class. Those don't exist though, and `instance` just referred to an instance of generic `HttpHeader` name and value.

In the Scala DSL, you can't extract a generic `HttpHeader` instance when providing a specific name, so you can't do that in Java anymore now either. It's also a bit nonsensical: if you provide an exact name for the header, you already know its name.

rkuhn commented:
Looked at the TODOs and FIXMEs and it looks like we are nearly there!

jypma committed:
!htt #18919 #19519 Align Java HTTP server DSL with Scala

This commits replaces the Java HTTP server DSL with a Java-8 centric one
which exposes all scala DSL concepts to be usable from Java, including
custom directives, (un)marshallers, rejections, headers, and type safety
for path and query parameters.

ktoso commented:
Hi @jypma, I'd like to merge this PR this week and add missing bits if there are any.
Could you go over docs and try address them too – seems like last missing bit (other than things that we'll add afterwards)?

ktoso commented:
Chatted with @jypma just now. He's away for a bit, so we'll need to take over this PR, I'll attempt that soon.

ktoso commented:
I'll take it over the finish line in #20088, thanks for your awesome work.
Please feel free to help out on the effort once you're back!
